core/src/rust/tantivy_utils/src/query/cache.rs

```diff
@@ -0,0 +1,245 @@
+//! Cached query support
+
+use std::{hash::Hash, sync::Arc};
+
+use quick_cache::{sync::Cache, Equivalent, Weighter};
+use tantivy::{
+    collector::SegmentCollector,
+    query::{EnableScoring, Query, Weight},
+    schema::{Field, Schema},
+    Searcher, SegmentId, TantivyError,
+};
+use tantivy_common::BitSet;
+
+use crate::collectors::limited_collector::{
+    LimitCounter, LimitedCollector, LimitedSegmentCollector,
+};
+
+use super::bitset_weight::BitSetWeight;
+
+/// Cache for query results
+pub struct QueryCache<QueryType, WeighterType>
+where
+    QueryType: CachableQuery,
+    WeighterType: Weighter<(SegmentId, QueryType), Arc<BitSet>> + Default + Clone,
+{
+    // Cache of query -> docs
+    cache: Cache<(SegmentId, QueryType), Arc<BitSet>, WeighterType>,
+}
+
+/// Trait for cachable query keys
+pub trait CachableQuery: Eq + PartialEq + Hash + Clone {
+    /// Should this query be cached?
+    fn should_cache(&self) -> bool;
+
+    fn to_query(
+        &self,
+        schema: &Schema,
+        default_field: Option<Field>,
+    ) -> Result<Box<dyn Query>, TantivyError>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CachableQueryKey<'a, QueryType>(pub SegmentId, pub &'a QueryType)
+where
+    QueryType: Clone + PartialEq + Eq;
+
+impl<'a, QueryType> From<CachableQueryKey<'a, QueryType>> for (SegmentId, QueryType)
+where
+    QueryType: Clone + PartialEq + Eq,
+{
+    fn from(value: CachableQueryKey<'a, QueryType>) -> Self {
+        (value.0, value.1.clone())
+    }
+}
+
+impl<'a, QueryType> Equivalent<(SegmentId, QueryType)> for CachableQueryKey<'a, QueryType>
+where
+    QueryType: Clone + PartialEq + Eq,
+{
+    fn equivalent(&self, key: &(SegmentId, QueryType)) -> bool {
+        self.0 == key.0 && *self.1 == key.1
+    }
+}
+
+// Tuning parameters for query cache
+const QUERY_CACHE_MAX_SIZE_BYTES: u64 = 50_000_000;
+// Rough estimate of bitset size - 250k docs
+const QUERY_CACHE_AVG_ITEM_SIZE: u64 = 31250;
```

Comment:
Can these be made configurable?

Reply:
Sure, will wire this up (and the others called out)